crypto/blake256/hasher.go

```diff
@@ -0,0 +1,414 @@
+//
+// The hasher may no longer be used after invoking this method.  Callers always
+// run finalize on a copy of the hasher so the original hasher state is not
+// modified.
+func (h *hasher) finalize224() [Size224]byte {
+	const lenPreambleBit = 0x00
+	h.finalize(lenPreambleBit)
+	return wordsToBytes224(h.cv)
+}
+
+// wordsToBytes256 converts an array of 8 32-bit unsigned big-endian words to an
+// array of 32 bytes.
+func wordsToBytes256(cv [8]uint32) (out [32]byte) {
+	binary.BigEndian.PutUint32(out[28:], cv[7])
+	binary.BigEndian.PutUint32(out[24:], cv[6])
+	binary.BigEndian.PutUint32(out[20:], cv[5])
+	binary.BigEndian.PutUint32(out[16:], cv[4])
+	binary.BigEndian.PutUint32(out[12:], cv[3])
+	binary.BigEndian.PutUint32(out[8:], cv[2])
+	binary.BigEndian.PutUint32(out[4:], cv[1])
+	binary.BigEndian.PutUint32(out[0:], cv[0])
+	return out
+}
+
+// finalize256 finalizes of the rolling hash by writing any remaining partial
+// block data and appending the necessary padding for BLAKE-256.
+//
+// The hasher may no longer be used after invoking this method.  Callers always
+// run finalize on a copy of the hasher so the original hasher state is not
+// modified.
+func (h *hasher) finalize256() [Size]byte {
+	const lenPreambleBit = 0x01
+	h.finalize(lenPreambleBit)
+	return wordsToBytes256(h.cv)
+}
+
+// putSavedState serializes the intermediate state directly into the passed byte
+// slice.  The target slice MUST have at least [SavedStateSize] bytes available
+// or it will panic.
+func (h *hasher) putSavedState(target []byte, prefix uint32) {
+	var offset uint32
+	binary.BigEndian.PutUint32(target[offset:], prefix)
+	offset += 4
+	for _, cv := range h.cv {
+		binary.BigEndian.PutUint32(target[offset:], cv)
+		offset += 4
+	}
+	for _, s := range h.s {
+		binary.BigEndian.PutUint32(target[offset:], s)
+		offset += 4
+	}
+	binary.BigEndian.PutUint64(target[offset:], h.count)
+	offset += 8
+	offset += uint32(copy(target[offset:], h.buf[:]))
+	binary.BigEndian.PutUint32(target[offset:], h.nbuf)
+}
+
+// saveState appends the current intermediate state of the rolling hash prefixed
+// by the passed value to the provided slice and returns the resulting slice.
+// It does not change the underlying hash state.
+//
+// The provided prefix is expected to either be [statePrefix224] or
+// [statePrefix256] depending on which hash variant is being saved.
+//
+// As described by the [hasher] documentation, the hasher instance can simply be
+// copied to achieve the same result much more efficiently when the caller is
+// able to keep a copy.  Therefore, that approach should be preferred when
+// possible.
+//
+// However, the ability to serialize the state is also provided to enable
+// sharing it across process boundaries.
+func (h *hasher) saveState(target []byte, prefix uint32) []byte {
+	// Create a new array to write into and append that when there is not enough
+	// space remining in the slice or write directly into it when there is.
```

Comment:
This sentence is hard to read due to the confusing mix of the words "that when there".

Suggested rewording:

"If the target slice doesn't have enough space remaining, create a new array and append it to the target. Otherwise, write directly into it."

Reply:
Thanks.  I reworded it to closely match your suggestion.  I don't like leading sentences with if so I reversed the order on that one, but otherwise it's your suggestion.

Reply:
New change reads well 👍